File: backend/app/routers/audit.py
"""Audit trail endpoints."""

from __future__ import annotations
import datetime

from fastapi import APIRouter, Depends, Query
from sqlalchemy import select, func
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.models import AuditLog
from app.schemas import AuditLogListResponse

router = APIRouter(prefix="/api/audit", tags=["audit"])


@router.get("", response_model=AuditLogListResponse)
async def list_audit_logs(
    job_id: int | None = Query(None, description="Filter by job ID"),
    action_type: str | None = Query(None, description="Filter by action type"),
    entity_type: str | None = Query(None, description="Filter by entity type"),
    from_date: datetime.datetime | None = Query(None, description="Start date (inclusive)"),
    to_date: datetime.datetime | None = Query(None, description="End date (inclusive)"),
    page: int = Query(1, ge=1),
    page_size: int = Query(50, ge=1, le=200),
    db: AsyncSession = Depends(get_db),
):
    """List audit log entries with optional filters."""

    query = select(AuditLog)

    if job_id is not None:
        query = query.where(AuditLog.job_id == job_id)
    if action_type:
        query = query.where(AuditLog.action_type == action_type)
    if entity_type:
        query = query.where(AuditLog.entity_type == entity_type)
    if from_date:
        query = query.where(AuditLog.timestamp >= from_date)
    if to_date:
        query = query.where(AuditLog.timestamp <= to_date)

    # Total count
    count_query = select(func.count()).select_from(query.subquery())
    total_result = await db.execute(count_query)
    total = total_result.scalar() or 0

    # Paginated results (newest first)
    query = query.order_by(AuditLog.timestamp.desc())
    query = query.offset((page - 1) * page_size).limit(page_size)

    result = await db.execute(query)
    items = result.scalars().all()

    return AuditLogListResponse(items=items, total=total)

File: frontend/src/lib/api/audit.js
import { apiFetch } from './client';

export async function listAuditLogs({
    jobId = null,
    actionType = null,
    entityType = null,
    fromDate = null,
    toDate = null,
    page = 1,
    pageSize = 50,
} = {}) {
    const params = new URLSearchParams({ page, page_size: pageSize });
    if (jobId !== null) params.set('job_id', jobId);
    if (actionType) params.set('action_type', actionType);
    if (entityType) params.set('entity_type', entityType);
    if (fromDate) params.set('from_date', fromDate);
    if (toDate) params.set('to_date', toDate);

    const res = await apiFetch(`/api/audit?${params}`);
    if (!res.ok) throw new Error('Failed to load audit logs');
    return res.json();
}
